dask/dataframe/io/parquet/arrow.py

```diff
@@ -1766,14 +1776,17 @@ def _read_table(
             for partition in partitions:
                 if partition.name not in arrow_table.schema.names:
                     # We read from file paths, so the partition
-                    # columns are NOT in our table yet.
+                    # columns may NOT be in our table yet.
                     cat = keys_dict.get(partition.name, None)
-                    cat_ind = np.full(
-                        len(arrow_table), partition.keys.get_loc(cat), dtype="i4"
-                    )
-                    arr = pa.DictionaryArray.from_arrays(
-                        cat_ind, pa.array(partition.keys)
-                    )
+                    if not len(partition.keys):
+                        arr = pa.array(np.full(len(arrow_table), cat))
```

Comment:
Non-categorical case.  Note that pyarrow will usually populate this column during the read. Therefore, the `if partition.name not in arrow_table.schema.names:` scope means that this line won't execute for newer pyarrow versions.